[1.21.1][Neoforge 21.1.68][19.0.23-beta] Guide Crashes Client

### Describe the bug

Opening the guidebook for more than a minute will eventually crash the client - timing and page seems to be random.  There's no crash-log, but there is a rendering error in the debug.log.

<details>
  <summary>debug.log</summary>

```
[26Oct2024 15:26:03.996] [Render thread/WARN] [oshi.util.platform.windows.WmiQueryHandler/]: COM exception querying Win32_Processor, which might not be on your system. Will not attempt to query it again. Error was -2147024882: Not enough memory resources are available to complete this operation.(HRESULT: 8007000e)
[26Oct2024 15:26:04.001] [Render thread/WARN] [oshi.util.platform.windows.WmiQueryHandler/]: COM exception querying Win32_Processor, which might not be on your system. Will not attempt to query it again. Error was -2147217402: (HRESULT: 80041006)
[26Oct2024 15:26:04.003] [Render thread/DEBUG] [oshi.util.FileUtil/]: No oshi.architecture.properties file found from ClassLoader cpw.mods.modlauncher.TransformingClassLoader@3cb04dd
[26Oct2024 15:26:04.007] [Render thread/WARN] [oshi.util.platform.windows.WmiQueryHandler/]: COM exception querying Win32_PhysicalMemory, which might not be on your system. Will not attempt to query it again. Error was -2147217402: (HRESULT: 80041006)
[26Oct2024 15:26:04.020] [Render thread/ERROR] [net.minecraft.client.Minecraft/FATAL]: Reported exception thrown!
net.minecraft.ReportedException: Rendering screen
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.client.renderer.GameRenderer.render(GameRenderer.java:1120) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.client.Minecraft.runTick(Minecraft.java:1195) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.client.Minecraft.run(Minecraft.java:807) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.client.main.Main.main(Main.java:230) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[?:?]
	at java.base/java.lang.reflect.Method.invoke(Method.java:580) ~[?:?]
	at MC-BOOTSTRAP/fml_loader@4.0.24/net.neoforged.fml.loading.targets.CommonLaunchHandler.runTarget(CommonLaunchHandler.java:136) ~[loader-4.0.24.jar%23130!/:4.0]
	at MC-BOOTSTRAP/fml_loader@4.0.24/net.neoforged.fml.loading.targets.CommonLaunchHandler.clientService(CommonLaunchHandler.java:124) ~[loader-4.0.24.jar%23130!/:4.0]
	at MC-BOOTSTRAP/fml_loader@4.0.24/net.neoforged.fml.loading.targets.CommonClientLaunchHandler.runService(CommonClientLaunchHandler.java:32) ~[loader-4.0.24.jar%23130!/:4.0]
	at MC-BOOTSTRAP/fml_loader@4.0.24/net.neoforged.fml.loading.targets.CommonLaunchHandler.lambda$launchService$4(CommonLaunchHandler.java:118) ~[loader-4.0.24.jar%23130!/:4.0]
	at MC-BOOTSTRAP/cpw.mods.modlauncher@11.0.4/cpw.mods.modlauncher.LaunchServiceHandlerDecorator.launch(LaunchServiceHandlerDecorator.java:30) [modlauncher-11.0.4.jar%23103!/:?]
	at MC-BOOTSTRAP/cpw.mods.modlauncher@11.0.4/cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:53) [modlauncher-11.0.4.jar%23103!/:?]
	at MC-BOOTSTRAP/cpw.mods.modlauncher@11.0.4/cpw.mods.modlauncher.LaunchServiceHandler.launch(LaunchServiceHandler.java:71) [modlauncher-11.0.4.jar%23103!/:?]
	at MC-BOOTSTRAP/cpw.mods.modlauncher@11.0.4/cpw.mods.modlauncher.Launcher.run(Launcher.java:103) [modlauncher-11.0.4.jar%23103!/:?]
	at MC-BOOTSTRAP/cpw.mods.modlauncher@11.0.4/cpw.mods.modlauncher.Launcher.main(Launcher.java:74) [modlauncher-11.0.4.jar%23103!/:?]
	at MC-BOOTSTRAP/cpw.mods.modlauncher@11.0.4/cpw.mods.modlauncher.BootstrapLaunchConsumer.accept(BootstrapLaunchConsumer.java:26) [modlauncher-11.0.4.jar%23103!/:?]
	at MC-BOOTSTRAP/cpw.mods.modlauncher@11.0.4/cpw.mods.modlauncher.BootstrapLaunchConsumer.accept(BootstrapLaunchConsumer.java:23) [modlauncher-11.0.4.jar%23103!/:?]
	at cpw.mods.bootstraplauncher@2.0.2/cpw.mods.bootstraplauncher.BootstrapLauncher.run(BootstrapLauncher.java:210) [bootstraplauncher-2.0.2.jar:?]
	at cpw.mods.bootstraplauncher@2.0.2/cpw.mods.bootstraplauncher.BootstrapLauncher.main(BootstrapLauncher.java:69) [bootstraplauncher-2.0.2.jar:?]
Caused by: java.lang.OutOfMemoryError: Failed to resize buffer from 262144 bytes to 524288 bytes
	at TRANSFORMER/minecraft@1.21.1/com.mojang.blaze3d.vertex.ByteBufferBuilder.resize(ByteBufferBuilder.java:55) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/com.mojang.blaze3d.vertex.ByteBufferBuilder.ensureCapacity(ByteBufferBuilder.java:47) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/com.mojang.blaze3d.vertex.ByteBufferBuilder.reserve(ByteBufferBuilder.java:38) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/com.mojang.blaze3d.vertex.MeshData$SortState.buildSortedIndexBuffer(MeshData.java:91) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/com.mojang.blaze3d.vertex.MeshData.sortQuads(MeshData.java:71) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.client.renderer.MultiBufferSource$BufferSource.endBatch(MultiBufferSource.java:94) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.client.renderer.MultiBufferSource$BufferSource.endBatch(MultiBufferSource.java:85) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.client.renderer.MultiBufferSource$BufferSource.getBuffer(MultiBufferSource.java:56) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.client.gui.Font$StringRenderOutput.accept(Font.java:413) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.util.StringDecomposer.feedChar(StringDecomposer.java:17) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.util.StringDecomposer.iterate(StringDecomposer.java:43) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.util.FormattedCharSequence.lambda$forward$2(FormattedCharSequence.java:23) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.client.gui.Font.renderText(Font.java:261) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.client.gui.Font.drawInternal(Font.java:223) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.client.gui.Font.drawInBatch(Font.java:128) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.client.gui.Font.drawInBatch(Font.java:111) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/ae2@19.0.23-beta/appeng.client.guidebook.render.RenderContext.renderTextInBatch(RenderContext.java:168) ~[appliedenergistics2-19.0.23-beta.jar%23569!/:?]
	at TRANSFORMER/ae2@19.0.23-beta/appeng.client.guidebook.layout.flow.LineTextRun.renderBatch(LineTextRun.java:23) ~[appliedenergistics2-19.0.23-beta.jar%23569!/:?]
	at TRANSFORMER/ae2@19.0.23-beta/appeng.client.guidebook.layout.flow.FlowBuilder.renderBatch(FlowBuilder.java:48) ~[appliedenergistics2-19.0.23-beta.jar%23569!/:?]
	at TRANSFORMER/ae2@19.0.23-beta/appeng.client.guidebook.document.block.LytParagraph.renderBatch(LytParagraph.java:92) ~[appliedenergistics2-19.0.23-beta.jar%23569!/:?]
	at TRANSFORMER/ae2@19.0.23-beta/appeng.client.guidebook.document.block.LytDocument.renderBatch(LytDocument.java:119) ~[appliedenergistics2-19.0.23-beta.jar%23569!/:?]
	at TRANSFORMER/ae2@19.0.23-beta/appeng.client.guidebook.screen.GuideScreen.render(GuideScreen.java:243) ~[appliedenergistics2-19.0.23-beta.jar%23569!/:?]
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.client.gui.screens.Screen.renderWithTooltip(Screen.java:112) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	at TRANSFORMER/neoforge@21.1.68/net.neoforged.neoforge.client.ClientHooks.drawScreenInternal(ClientHooks.java:428) ~[neoforge-21.1.68-universal.jar%23539!/:?]
	at TRANSFORMER/neoforge@21.1.68/net.neoforged.neoforge.client.ClientHooks.drawScreen(ClientHooks.java:422) ~[neoforge-21.1.68-universal.jar%23539!/:?]
	at TRANSFORMER/minecraft@1.21.1/net.minecraft.client.renderer.GameRenderer.render(GameRenderer.java:1092) ~[client-1.21.1-20240808.144430-srg.jar%23538!/:?]
	... 18 more
```

</details>



### How to reproduce the bug

Browse the guidebook. The exact page is and timing is random.

### Expected behavior

Client doesn't crash

### Additional details

System details:
Ryzen 9 7950X
Radeon 7900XTX
Windows 11
8GB java memory out of 32GB RAM

### Which minecraft version are you using?

1.21

### On which mod loaders does it happen?

NeoForge

### Crash log

https://pastebin.com/UER0sQCq

--- a/src/main/java/guideme/internal/screen/GuideScreen.java
+++ b/src/main/java/guideme/internal/screen/GuideScreen.java
@@ -1,7 +1,6 @@
 package guideme.internal.screen;
 
 import com.mojang.blaze3d.systems.RenderSystem;
-import com.mojang.blaze3d.vertex.ByteBufferBuilder;
 import guideme.Guide;
 import guideme.GuidePage;
 import guideme.PageAnchor;
@@ -50,7 +49,6 @@ import net.minecraft.client.gui.components.Button;
 import net.minecraft.client.gui.screens.ConfirmLinkScreen;
 import net.minecraft.client.gui.screens.Screen;
 import net.minecraft.client.gui.screens.inventory.tooltip.TooltipRenderUtil;
-import net.minecraft.client.renderer.MultiBufferSource;
 import net.minecraft.network.chat.Component;
 import net.minecraft.resources.ResourceLocation;
 import net.neoforged.fml.ModContainer;
@@ -253,7 +251,7 @@ public class GuideScreen extends Screen implements GuideUiHost {
 
         renderTitle(documentRect, context);
 
-        renderExternalpageSource(documentRect, context);
+        renderExternalPageSource(documentRect, context);
 
         super.render(guiGraphics, mouseX, mouseY, partialTick);
 
@@ -266,7 +264,7 @@ public class GuideScreen extends Screen implements GuideUiHost {
 
     }
 
-    private void renderExternalpageSource(LytRect documentRect, SimpleRenderContext context) {
+    private void renderExternalPageSource(LytRect documentRect, SimpleRenderContext context) {
         // Render the source of the content
         var externalSource = getExternalSourceName();
         if (externalSource != null) {
@@ -792,7 +790,7 @@ public class GuideScreen extends Screen implements GuideUiHost {
         }
 
         var poseStack = guiGraphics.pose();
-        var bufferSource = MultiBufferSource.immediate(new ByteBufferBuilder(512));
+        var bufferSource = Minecraft.getInstance().renderBuffers().bufferSource();
         poseStack.pushPose();
         poseStack.translate(0.0, 0.0, zOffset);
         int currentY = y;

--- a/src/main/java/guideme/render/RenderContext.java
+++ b/src/main/java/guideme/render/RenderContext.java
@@ -1,6 +1,5 @@
 package guideme.render;
 
-import com.mojang.blaze3d.vertex.ByteBufferBuilder;
 import com.mojang.blaze3d.vertex.PoseStack;
 import guideme.color.ColorValue;
 import guideme.color.ConstantColor;
@@ -157,9 +156,9 @@ public interface RenderContext {
     }
 
     default void renderText(String text, ResolvedTextStyle style, float x, float y) {
-        var bufferSource = MultiBufferSource.immediate(new ByteBufferBuilder(512));
+        var bufferSource = beginBatch();
         renderTextInBatch(text, style, x, y, bufferSource);
-        bufferSource.endBatch();
+        endBatch(bufferSource);
     }
 
     default void renderTextInBatch(String text, ResolvedTextStyle style, float x, float y, MultiBufferSource buffers) {
@@ -210,7 +209,7 @@ public interface RenderContext {
     }
 
     default MultiBufferSource.BufferSource beginBatch() {
-        return MultiBufferSource.immediate(new ByteBufferBuilder(512));
+        return Minecraft.getInstance().renderBuffers().bufferSource();
     }
 
     default void endBatch(MultiBufferSource.BufferSource batch) {

